### PR TITLE
Feature/legalhold

### PR DESCRIPTION
**Pull request by tiago-loureiro**
To be unwip'ed soon (and description updated)

### CODE REVIEW

File: services/brig/src/Brig/API/Client.hs [hunk 1]
```diff
@@ -46,24 +48,23 @@ import qualified Brig.User.Auth.Cookie as Auth
 
 -- nb. We must ensure that the set of clients known to brig is always
 -- a superset of the clients known to galley.
-addClient :: UserId -> ConnId -> Maybe IP -> NewClient -> ExceptT ClientError AppIO Client
+addClient :: UserId -> Maybe ConnId -> Maybe IP -> NewClient -> ExceptT ClientError AppIO Client
```
Comment (tiago-loureiro) [inline services/brig/src/Brig/API/Client.hs]: Not sure why `ConnId` is now a `Maybe`, was this meant to be just temporary... ?

File: services/brig/src/Brig/API/Client.hs [hunk 2]
```diff
@@ -46,24 +48,23 @@ import qualified Brig.User.Auth.Cookie as Auth
...
-    (clt, old, count) <- Data.addClient u newId new loc !>> ClientDataError
+    (clt, old, count) <- Data.addClient u clientId' new loc !>> ClientDataError
     let usr = accountUser acc
     lift $ do
-        for_ old $ execDelete u (Just con)
+        for_ old $ execDelete u con
         Intra.newClient u (clientId clt)
-        Intra.onClientEvent u (Just con) (ClientAdded u clt)
+        Intra.onClientEvent u con (ClientAdded u clt)
+        when (clientType clt == LegalHoldClientType) $ Intra.onUserEvent u con (UserLegalHoldEnabled u)
```
Comment (tiago-loureiro) [inline services/brig/src/Brig/API/Client.hs]: Should eventually be done elsewhere, see `-- TODO: send event at this point`

File: libs/brig-types/src/Brig/Types/Team/LegalHold.hs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+-- Request body definition for the @/confirm@ endpoint on the LegalHold Service
+data LegalHoldServiceConfirm =
+    LegalHoldServiceConfirm
+    { lhcClientId     :: !ClientId
+    , lhcUserId       :: !UserId
+    , lhcTeamId       :: !TeamId
+    , lhcRefreshToken :: !Text -- ^ Replace with Legal Hold Token Type
+    } deriving stock (Eq, Show, Generic)
+
+-- TODO: Review this in a separate PR, requires sync'ing with LH team
```
Comment (tiago-loureiro) [inline libs/brig-types/src/Brig/Types/Team/LegalHold.hs]: @dkovacevic can you verify that this looks good ?

File: libs/brig-types/src/Brig/Types/Team/LegalHold.hs [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+instance FromJSON LegalHoldServiceConfirm where
+  parseJSON = withObject "LegalHoldServiceConfirm" $ \o ->
+    LegalHoldServiceConfirm
+        <$> o .: "client_id"
+        <*> o .: "user_id"
+        <*> o .: "team_id"
+        <*> o .: "refresh_token"
+
+-- TODO: Review this in a separate PR, requires sync'ing with LH team
```
Comment (tiago-loureiro) [inline libs/brig-types/src/Brig/Types/Team/LegalHold.hs]: @dkovacevic can you verify that this looks good ?

File: services/galley/src/Galley/API/LegalHold.hs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        <- fromJsonBody req
+
+    (key :: ServiceKey, fpr :: Fingerprint Rsa)
+        <- LHService.validateServiceKey (newLegalHoldServiceKey newService)
+               >>= maybe (throwM legalHoldServiceInvalidKey) pure
+    LHService.checkLegalHoldServiceStatus fpr (newLegalHoldServiceUrl newService)
+
+    let service = legalHoldService tid fpr newService key
+    LegalHoldData.createSettings service
+    -- TODO: Refactor and use `json` function instead which ensures correct content-type is set, etc.
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/API/LegalHold.hs]: Looks easy enough to fix now

File: libs/brig-types/src/Brig/Types/Team/LegalHold.hs (hunk 3 above)
Comment (dkovacevic) [reply libs/brig-types/src/Brig/Types/Team/LegalHold.hs]: I am not sure I understand this code right but LH documentation is here:
https://services.zinfra.io/legalhold/swagger#!/default/confirm

File: services/galley/src/Galley/API/LegalHold.hs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+    mPreKeys <- LegalHoldData.selectPendingPrekeys uid
+    (prekeys, lastPrekey') <- case mPreKeys of
+        Nothing -> do
+            Log.info $ Log.msg @Text "No prekeys found"
+            throwM noLegalHoldDeviceAllocated
+        Just keys -> pure keys
+
+    clientId <- Client.addLegalHoldClientToUser uid connId prekeys lastPrekey'
+    -- TODO: this may be redundant
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/API/LegalHold.hs]: This _is_ redundant, please remove @fisx

File: services/galley/src/Galley/API/Swagger.hs [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+      where
+        opts = defaultSchemaOptions
+          { fieldLabelModifier = \case
+              "newLegalHoldClientPrekeys"     -> "prekeys"
+              "newLegalHoldClientLastKey"     -> "last_prekey"
+          }
+
+-- | this type is only introduce locally here to generate the schema for 'UserLegalHoldStatusResponse'.
+--   TODO: Use SimpleClientId in the UserLegalHoldStatusResponse datatype or re-use PubClient datatype
+--         which would greatly simplify these instances
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/API/Swagger.hs]: @fisx feel free to refactor this before unwipping. Also, the naming could be different 🙈

File: services/galley/src/Galley/External/LegalHoldService.hs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+      . Bilge.method POST
+      . Bilge.acceptJson
+      . Bilge.expect2xx
+
+-- | @POST /confirm@
+-- Confirm that a device has been linked to a user and provide an authorization token
+confirmLegalHold :: ClientId
+                 -> TeamId
+                 -> UserId
+                 -> OpaqueAuthToken -- ^ TODO: Replace with 'LegalHold' token type
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/External/LegalHoldService.hs]: Being worked on [here](https://github.com/wireapp/wire-server/pull/761)

File: services/galley/src/Galley/External/LegalHoldService.hs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+removeLegalHold :: TeamId
+                 -> UserId
+                 -> Galley ()
+removeLegalHold tid uid = do
+    void $ makeLegalHoldServiceRequest tid reqParams
+  where
+    reqParams =
+        -- TODO: Currently this OVERWRITES any path on the base URL;
+        -- We should come up with a better solution.
+        Bilge.paths ["remove"]
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/External/LegalHoldService.hs]: We should definitely fix this before merging; this path should be appended to `legalHoldServiceUrl`

File: services/galley/src/Galley/External/LegalHoldService.hs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+confirmLegalHold :: ClientId
+                 -> TeamId
+                 -> UserId
+                 -> OpaqueAuthToken -- ^ TODO: Replace with 'LegalHold' token type
+                 -> Galley ()
+confirmLegalHold clientId tid uid legalHoldAuthToken = do
+    void $ makeLegalHoldServiceRequest tid reqParams
+  where
+    reqParams =
+        Bilge.paths ["confirm"]
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/External/LegalHoldService.hs]: We should definitely fix this before merging; this path should be appended to `legalHoldServiceUrl`

File: services/brig/test/resources/turn/servers-v2.txt [hunk 11]
```diff
@@ -1,2 +1 @@
-turn:localhost:3478
-turn:localhost:3479
+turn:localhost:3478
```
Comment (fisx) [inline services/brig/test/resources/turn/servers-v2.txt]: this keeps getting committed, but it really shouldn't...

File: services/galley/src/Galley/API/LegalHold.hs (hunk 6 above)
Comment (fisx) [reply services/galley/src/Galley/API/LegalHold.hs]: 5325b1350534f1069cd07fcb893e78873a54d91c

File: services/galley/src/Galley/External/LegalHoldService.hs (hunk 10 above)
Comment (fisx) [reply services/galley/src/Galley/External/LegalHoldService.hs]: isn't this already happening?

- https://github.com/wireapp/wire-server/blob/5325b1350534f1069cd07fcb893e78873a54d91c/services/galley/test/integration/API/Teams/LegalHold.hs#L743
- https://github.com/wireapp/wire-server/blob/5325b1350534f1069cd07fcb893e78873a54d91c/services/galley/test/integration/API/Teams/LegalHold.hs#L780-L786
- https://github.com/wireapp/wire-server/blob/5325b1350534f1069cd07fcb893e78873a54d91c/services/galley/src/Galley/External/LegalHoldService.hs#L176

File: services/galley/src/Galley/External/LegalHoldService.hs (hunk 9 above)
Comment (fisx) [reply services/galley/src/Galley/External/LegalHoldService.hs]: see above

File: services/brig/src/Brig/API/Client.hs (hunk 1 above)
Comment (fisx) [reply services/brig/src/Brig/API/Client.hs]: Do we always have a connId for LH devices?  Do we need one before the device is approved, but can't have one then?

Not sure...

File: services/galley/src/Galley/API/LegalHold.hs (hunk 5 above)
Comment (fisx) [reply services/galley/src/Galley/API/LegalHold.hs]: 9b289db9497337811a8a86316c05907b9e4c2163

(it was easy, only took me two commits...  :-))

File: libs/brig-types/src/Brig/Types/Team/LegalHold.hs (hunk 3 above)
Comment (fisx) [reply libs/brig-types/src/Brig/Types/Team/LegalHold.hs]: @dkovacevic or is your swagger out of date?

**Comment by fisx** [comment]
looks like a ci hickup.  i restarted the integration tests.

File: services/galley/src/Galley/External/LegalHoldService.hs (hunk 10 above)
Comment (tiago-loureiro) [reply services/galley/src/Galley/External/LegalHoldService.hs]: Oh so the comment was outdated?

Comment (fisx) [reply services/galley/src/Galley/External/LegalHoldService.hs]: Seems so, yes.  Sorry about that.

Comment (fisx) [reply services/galley/src/Galley/External/LegalHoldService.hs]: (I only saw the comment after I answered to your answer to it.  I think your answer was to the line above the comment, so it didn't show in the email.  Life is full of tricks!)

**Comment by jschaul** [comment]
This large PR/feature could really use some documentation under `docs/reference` before merging, now that the concepts are still fresh-ish.

**Pull request merged**